### ISSUE TITLE
experimentalObjectRestSpread should allow AssignmentExpression

### ISSUE THREAD

**Issue by gre** [eslint/espree#167]
I have first created the issue at https://github.com/eslint/eslint/issues/3030

```js
const foo = {
  a: 1
}
const obj = {
  foo
};
const bar = "bar";
const res2 = {
  ...obj.foo, // does not get parsed, unexpected '.'
  bar
};
```

**Comment by michaelficarra** [comment]
Note that AssignmentExpression is only allowed in spread position. Object rest, like array and parameter list rest, still only allows BindingIdentifier.

**Comment by nzakas** [comment]
Thanks, I'll take a look.

**Comment by nzakas** [comment]
Working on this.

**Comment by gre** [comment]
Hi,
thanks for the fast handling of the issue :)
I still can't make this working (I'm not having the same error though), I don't know if this is from espree or eslint.

Here is what happens to me:

```
$ npm install -g eslint@1.0.0-rc-1     # this grabs espree 2.2.2
...
$ eslint test.js
/Users/gre/.nvm/versions/io.js/v2.3.4/lib/node_modules/eslint/lib/rules/no-dupe-keys.js:25
                var keyName = property.key.name || property.key.value,
                                          ^
TypeError: Cannot read property 'name' of undefined
    at /Users/gre/.nvm/versions/io.js/v2.3.4/lib/node_modules/eslint/lib/rules/no-dupe-keys.js:25:43
    at Array.forEach (native)
    at EventEmitter.ObjectExpression (/Users/gre/.nvm/versions/io.js/v2.3.4/lib/node_modules/eslint/lib/rules/no-dupe-keys.js:24:29)
    at emitOne (events.js:82:20)
    at EventEmitter.emit (events.js:169:7)
    at Controller.controller.traverse.enter (/Users/gre/.nvm/versions/io.js/v2.3.4/lib/node_modules/eslint/lib/eslint.js:790:25)
    at Controller.__execute (/Users/gre/.nvm/versions/io.js/v2.3.4/lib/node_modules/eslint/node_modules/estraverse/estraverse.js:397:31)
    at Controller.traverse (/Users/gre/.nvm/versions/io.js/v2.3.4/lib/node_modules/eslint/node_modules/estraverse/estraverse.js:495:28)
    at EventEmitter.module.exports.api.verify (/Users/gre/.nvm/versions/io.js/v2.3.4/lib/node_modules/eslint/lib/eslint.js:783:24)
    at processText (/Users/gre/.nvm/versions/io.js/v2.3.4/lib/node_modules/eslint/lib/cli-engine.js:199:27)
```

**test.js**
```js
const a = { foo: {} };

const b = {
  ...a.foo,
  a
};
```

**.eslintrc**
```
{
    "rules": {
        "indent": [
            2,
            4
        ],
        "quotes": [
            2,
            "double"
        ],
        "linebreak-style": [
            2,
            "unix"
        ],
        "semi": [
            2,
            "always"
        ]
    },
    "env": {
        "es6": true,
        "browser": true
    },
    "extends": "eslint:recommended",
    "ecmaFeatures": {
      "jsx": true,
      "experimentalObjectRestSpread": true
    }
}
```

**Comment by gre** [comment]
ok I'll create it on eslint project, this seems to be related to no-dupe-keys
